core/consensus/pow/src/lib.rs

```diff
@@ -0,0 +1,399 @@
+// Copyright 2017-2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Proof of work consensus for Substrate.
+//!
+//! To use this engine, you can either implement a standalone `PowAlgorithm`
+//! or use the runtime PoW algorithm engine. The runtime engine requires you
+//! to implement the `PowApi` defined in `substrate-consensus-pow-primitives`.
+//! There is also a helper module in SRML (`srml-pow`) to make the process easier.
+
+use std::sync::Arc;
+use std::thread;
+use client::{
+	BlockOf, blockchain::{HeaderBackend, ProvideCache},
+	block_builder::api::BlockBuilder as BlockBuilderApi, backend::AuxStore,
+};
+use sr_primitives::Justification;
+use sr_primitives::generic::{BlockId, Digest, DigestItem};
+use sr_primitives::traits::{Block as BlockT, Header as HeaderT, ProvideRuntimeApi};
+use srml_timestamp::{TimestampInherentData, InherentError as TIError};
+use pow_primitives::{Difficulty, Seal, POW_ENGINE_ID};
+use primitives::H256;
+use inherents::{InherentDataProviders, InherentData};
+use consensus_common::{
+	BlockImportParams, BlockOrigin, ForkChoiceStrategy,
+	well_known_cache_keys::Id as CacheKeyId, Environment, Proposer,
+};
+use consensus_common::import_queue::{BoxBlockImport, BasicQueue, Verifier};
+use codec::{Encode, Decode};
+use log::*;
+
+/// Auxiliary storage prefix for PoW engine.
+pub const POW_AUX_PREFIX: [u8; 4] = *b"PoW:";
+
+fn aux_key(hash: &H256) -> Vec<u8> {
+	POW_AUX_PREFIX.iter().chain(&hash[..])
+		.cloned().collect::<Vec<_>>()
+}
+
+/// Auxiliary storage data for PoW.
+#[derive(Encode, Decode, Clone, Debug, Default)]
+pub struct PowAux {
+	/// Total difficulty.
+	pub total_difficulty: Difficulty,
+}
+
+impl PowAux {
+	/// Read the auxiliary from client.
+	pub fn read<C: AuxStore>(client: &C, hash: &H256) -> Result<Self, String> {
+		let key = aux_key(hash);
+
+		match client.get_aux(&key).map_err(|e| format!("{:?}", e))? {
+			Some(bytes) => PowAux::decode(&mut &bytes[..]).map_err(|e| format!("{:?}", e)),
+			None => Ok(Default::default()),
+		}
+	}
+}
+
+/// Algorithm used for proof of work.
+pub trait PowAlgorithm<B: BlockT> {
+	/// Get the next block's difficulty.
+	fn difficulty(&self, parent: &BlockId<B>) -> Result<Difficulty, String>;
```

Comment:
How would this work for eg. Nakamoto PoW where block difficulty depends on the last *2016* blocks?

Reply:
The struct that implements `PowAlgorithm` is expected to take a reference to client `Arc<C>`. That will give it access to all past blocks, where it can find out their difficulty through auxiliary store. This way we can implement Nakamoto PoW. :)

Reply:
Great, thanks!

Reply:
But what about light clients? Is aux store available for them? I thought that they have access to headers only.

Reply:
As far as I understand light client should have access to aux store. Even if they don't, because we only need to store tiny amount of metadata in aux store, it's also quite easy to move that elsewhere.